test/hashing.spec.js

```diff
@@ -0,0 +1,1347 @@
+  });
+
+  it('Should remove qualifier and custom data attributes', () => {
+    expect(dlNormalizer('https://example.org/01/9780345418913/21/765tz?11=221109')).to.be.equal('https://id.gs1.org/01/09780345418913/21/765tz');
+    expect(dlNormalizer('https://example.org/01/9780345418913/21/765tz?abc=211121')).to.be.equal('https://id.gs1.org/01/09780345418913/21/765tz');
+  });
+  it('Should replace other gs1 keys', () => {
+    expect(dlNormalizer('https://id.gs1.org/00/340123453111111115')).to.be.equal('https://id.gs1.org/00/340123453111111115');
+    expect(dlNormalizer('https://id.gs1.org/414/4226350800008')).to.be.equal('https://id.gs1.org/414/4226350800008');
+    expect(dlNormalizer('https://example.co.uk/party/4226350800008')).to.be.equal('https://id.gs1.org/417/4226350800008');
+    expect(dlNormalizer('https://id.gs1.org/414/4280000000002/254/12')).to.be.equal('https://id.gs1.org/414/4280000000002/254/12');
+    expect(dlNormalizer('https://id.gs1.org/8003/03870585000552987')).to.be.equal('https://id.gs1.org/8003/03870585000552987');
+    expect(dlNormalizer('https://id.gs1.org/8004/0180451111ABC987')).to.be.equal('https://id.gs1.org/8004/0180451111ABC987');
+    expect(dlNormalizer('https://id.gs1.org/8018/385888700111111111')).to.be.equal('https://id.gs1.org/8018/385888700111111111');
+    expect(dlNormalizer('https://id.gs1.org/8017/440018922222222226')).to.be.equal('https://id.gs1.org/8017/440018922222222226');
+    expect(dlNormalizer('https://id.gs1.org/253/4602443000331XYZ')).to.be.equal('https://id.gs1.org/253/4602443000331XYZ');
+    expect(dlNormalizer('https://id.gs1.org/8010/0628165987/8011/9876')).to.be.equal('https://id.gs1.org/8010/0628165987/8011/9876');
+    expect(dlNormalizer('https://id.gs1.org/255/0811625999996554433')).to.be.equal('https://id.gs1.org/255/0811625999996554433');
+    expect(dlNormalizer('http://fashion.com/itip/040123451234560102/ser/ABC145')).to.be.equal('https://id.gs1.org/8006/040123451234560102/21/ABC145');
+  });
+  it('Should not replace non existent key ai', () => {
+    expect(dlNormalizer('http://example.org/123/4012345ABC987/456/4711')).to.be.equal('http://example.org/123/4012345ABC987/456/4711');
+    expect(dlNormalizer('http://id.gs1.org/99/9780345418913')).to.be.equal('http://id.gs1.org/99/9780345418913');
+    expect(dlNormalizer('http://id.gs1.org/01/9780345418913/99/utfgf')).to.be.equal('http://id.gs1.org/01/9780345418913/99/utfgf');
+  });
+  it('Should not replace ordinary web uri', () => {
+    expect(dlNormalizer('http://example.org/123/4012345ABC987/456/4711')).to.be.equal('http://example.org/123/4012345ABC987/456/4711');
+    expect(dlNormalizer('http://id.gs1.org/99/9780345418913')).to.be.equal('http://id.gs1.org/99/9780345418913');
+    expect(dlNormalizer('http://id.gs1.org/01/9780345418913/99/utfgf')).to.be.equal('http://id.gs1.org/01/9780345418913/99/utfgf');
+  });
+});
+
+describe('rule tests', () => {
+  it('should follow rule n°3', () => {
```

Comment:
Either describe those rules succinctly here, or provide a reference where the rule numbers can be looked up to understand what they require. Both here, and in code.